[event-hubs] Created a v2 -> v5 migration guide (#6501)
Azure/azure-sdk-for-js: sdk/eventhub/event-hubs/migrationguide.md

# Migration Guide (EventHubs v2 to v5)

> This document is intended for users that are familiar with EventHubs V2 (`@azure/event-hubs@2.x.x` & `@azure/event-processor-host@2.x.x`) and wish 

Reviewer: ```suggestion
This document is intended for users that are familiar with V2 of the JavaScript SDK for Event Hubs library (`@azure/event-hubs@2.x.x` & `@azure/event-processor-host@2.x.x`) and wish 
```

> to migrate their application to EventHubs V5.

Reviewer: Since v5 is for the library and not the service :)

```suggestion
to migrate their application to V5 of the same library.
```

For a more general introduction to Event Hubs, see the [readme.md](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/README.md) 

> for the @azure/event-hubs package.

Reviewer: If this line intended to re-direct new users, I would suggest to change it to

> For users new to the JavaScript SDK for Event Hubs, please see the [readme file for the @azure/event-hubs package](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/README.md)

Author: I added a period for @chradek but that's a great idea.

## General changes

In the interest of simplifying the API surface we've made two distinct
clients, rather than having a single `EventHubClient`:
* [EventHubProducerClient](https://docs.microsoft.com/en-us/javascript/api/@azure/event-hubs/eventhubproducerclient?view=azure-node-preview)
  for sending messages.
* [EventHubConsumerClient](https://docs.microsoft.com/en-us/javascript/api/@azure/event-hubs/eventhubconsumerclient?view=azure-node-preview) 
  for receiving messages.

> We've also merged the functionality from `EventProcessorHost` into 

Reviewer: ```suggestion
We've also merged the functionality from `EventProcessorHost` in the `@azure/event-processor-host` package into 
```

### Client constructors

| In v2                                          | Equivalent in v5                                                 | Sample |
|------------------------------------------------|------------------------------------------------------------------|--------|
| `EventHubClient.createFromConnectionString()`    | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts),  [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |
| `EventHubClient.createFromAadTokenCredentials()` | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [usingAadAuth](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/usingAadAuth.ts)

### Receiving events 

| In v2                                          | Equivalent in v5                                                 | Sample |
|------------------------------------------------|------------------------------------------------------------------|--------|

> | `EventHubClient.receive()` and `EventHubClient.receiveBatch()`                       | `EventHubConsumerClient.subscribe()`                               | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) |

Reviewer: An entry for `EventPosition.fromStart()` and `EventPosition.fromEnd()`?

Author: Added in a table for "Minor renames"

> | `EventProcessorHost.createFromConnectionString()`                           | `new EventHubConsumerClient(..., checkpointStore)`               | [receiveEventsUsingCheckpointStore](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEventsUsingCheckpointStore.ts) |

Reviewer: should this line for EPH move into the previous table of "Client constructors"?

Author: Makes sense to me :)

for a sample program demonstrating this.

### Migrating code from `EventHubClient` to `EventHubProducerClient` for sending events

In V2, there were multiple options on how to send data.

In V5, this has been consolidated into a more efficient `sendBatch` method. 
Batching merges information from multiple messages into a single send, reducing
the amount of network communication needed vs sending messages one at a time.

So in V2:
```typescript
const eventsToSend = [
  // events go here
];

const client = EventHubClient.createFromConnectionString(connectionString);

// Would fail if the total size of events exceed the max size supported by the library.
await client.sendBatch(eventsToSend, partitionId);
```

In V5:
```typescript
const producer = new EventHubProducerClient(connectionString);

const eventsToSend = [
  // events go here
];

let batch = await producer.createBatch();
let i = 0;

while (i < eventsToSend.length) {
  // messages can fail to be added to the batch if they exceed the maximum size configured for
  // the EventHub.
  const isAdded = batch.tryAdd(eventsToSend[i]);

  if (isAdded) {

>     console.log(`Added eventsToSend[${i}]`);

Reviewer: ```suggestion
    console.log(`Added event number ${i} to the batch`);
```

It would be good to match this up with a `console.log(`Creating a new batch`)` for everytime we create a new batch.
Then the logs would show a nice story of batches being created, filled and sent

    // is too large, even when it would be the _only_ message in the batch.
    //
    // At this point you'll need to decide if you're okay with skipping this message entirely
    // or find some way to shrink it.
    console.log(`Message was too large and can't be sent until it's made smaller. Skipping...`);
    ++i;
    continue;
  }

  // otherwise this just signals a good spot to send our batch
  console.log(`Batch is full - sending ${batch.count} messages as a single batch.`);
  await producer.sendBatch(batch);

  // and create a new one to house the next set of messages
  batch = await producer.createBatch();
}

// send any remaining messages, if any.
if (batch.count > 0) {
  console.log(`Sending remaining ${batch.count} messages as a single batch.`)
  await producer.sendBatch(batch);
}
```

### Migrating code from `EventProcessorHost` to `EventHubConsumerClient` for receiving events

In V2, `EventProcessorHost` allowed you to balance the load between multiple instances of your program when receiving events.

In V5, `EventHubConsumerClient` allows you to do the same with `subscribe()` and the 

> `SubscriptionEventHandlers` interface.

Reviewer: ```suggestion
`SubscriptionEventHandlers` interface as long as you pass in a checkpoint store.
```

Author: Went with slightly different wording (to mention you pass the checkpoint store into the constructor).

# Migration Guide (EventHubs v2 to v5)

This document is intended for users that are familiar with V2 of the JavaScript SDK for Event Hubs library (`@azure/event-hubs@2.x.x` & `@azure/event-processor-host@2.x.x`) and wish 
to migrate their application to V5 of the same library.

For users new to the JavaScript SDK for Event Hubs, please see the [readme file for the @azure/event-hubs package](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/README.md).

## General changes

In the interest of simplifying the API surface we've made two distinct
clients, rather than having a single `EventHubClient`:
* [EventHubProducerClient](https://docs.microsoft.com/en-us/javascript/api/@azure/event-hubs/eventhubproducerclient?view=azure-node-preview)
  for sending messages.
* [EventHubConsumerClient](https://docs.microsoft.com/en-us/javascript/api/@azure/event-hubs/eventhubconsumerclient?view=azure-node-preview) 
  for receiving messages.

We've also merged the functionality from `EventProcessorHost` in the `@azure/event-processor-host` package into 

> `EventHubConsumerClient`, allowing `EventHubConsumerClient` to be the single

Reviewer: ```suggestion
`EventHubConsumerClient` in the `@azure/event-hubs` package, allowing `EventHubConsumerClient` to be the single
```

    // is too large, even when it would be the _only_ message in the batch.
    //
    // At this point you'll need to decide if you're okay with skipping this message entirely
    // or find some way to shrink it.
    console.log(`Message was too large and can't be sent until it's made smaller. Skipping...`);
    ++i;
    continue;
  }

  // otherwise this just signals a good spot to send our batch
  console.log(`Batch is full - sending ${batch.count} messages as a single batch.`);
  await producer.sendBatch(batch);

  // and create a new one to house the next set of messages
  batch = await producer.createBatch();
}

// send any remaining messages, if any.
if (batch.count > 0) {
  console.log(`Sending remaining ${batch.count} messages as a single batch.`)
  await producer.sendBatch(batch);
}
```

### Migrating code from `EventProcessorHost` to `EventHubConsumerClient` for receiving events

In V2, `EventProcessorHost` allowed you to balance the load between multiple instances of your program when receiving events.

In V5, `EventHubConsumerClient` allows you to do the same with `subscribe()` and the 

> `SubscriptionEventHandlers` interface in addition to passing a `CheckpointStore` into the constructor.

Reviewer: I wonder if we should mention the `SubscriptionEventHandlers` interface here at all.
We don't end up referencing this interface in the sample, we just pass the callbacks.

How about

> In V5, `EventHubConsumerClient` allows you to do the same with the `subscribe()` method if you
pass a `CheckpointStore` to the constructor.

      console.log("[%s] Error: %O", ephName, error);
    }
  }
);

const onMessage: OnReceivedMessage = async (context: PartitionContext, event: EventData) => { }

const onError: OnReceivedError = (error: any) => {
  console.log("[%s] Received Error: %O", ephName, error);
};

await eph.start(onMessage, onError);
```

And in V5:
```typescript
import { EventHubConsumerClient, CheckpointStore } from "@azure/event-hubs";
import { ContainerClient } from "@azure/storage-blob";
import { BlobCheckpointStore } from "@azure/eventhubs-checkpointstore-blob";

const containerClient = new ContainerClient(storageConnectionString, storageContainerName);
const checkpointStore : CheckpointStore = new BlobCheckpointStore(containerClient);
const eventHubConsumerClient = new EventHubConsumerClient(consumerGroupName, ehConnectionString, eventHubName);

const subscription = eventHubConsumerClient.subscribe(
  partitionId, {
    // In V5 we deliver messages in batches, rather than a single message 
    // at a time. You can control the batch size via the options passed to the client.
    processEvents: (messages, context) => {},

    // Prior to V5 errors were handled by separate callbacks depending 

>     // on where they were thrown.

Reviewer: ```suggestion
    // on where they were thrown. When managing different partitions vs receiving from each partition.
```
